File: frontend/src/extensions/shared/modelPaths.js
const pathMapCache = new Map();
const pathMapRequests = new Map();

export function getModelPath(folder, filename) {
  if (!filename || filename === "None") {
    return "";
  }

  return pathMapCache.get(folder)?.[filename] ?? filename;
}

export async function loadModelPathMap(folder) {
  if (pathMapRequests.has(folder)) {
    return pathMapRequests.get(folder);
  }

  const request = fetch(`/alice-tools/model-paths/${encodeURIComponent(folder)}`)
    .then((response) => (response.ok ? response.json() : { paths: {} }))
    .then((data) => {
      const paths = data?.paths ?? {};
      pathMapCache.set(folder, paths);
      return paths;
    })
    .catch(() => {
      pathMapCache.set(folder, {});
      return {};
    });

  pathMapRequests.set(folder, request);
  return request;
}

File: py/model_paths.py
from aiohttp import web
from server import PromptServer
import folder_paths

MODEL_PATH_FOLDERS = {"loras", "controlnet"}


def build_model_path_map(folder):
    paths = {}
    for filename in folder_paths.get_filename_list(folder):
        full_path = folder_paths.get_full_path(folder, filename)
        if full_path:
            paths[filename] = full_path
    return paths


@PromptServer.instance.routes.get("/alice-tools/model-paths/{folder}")
async def get_model_paths(request):
    folder = request.match_info["folder"]
    if folder not in MODEL_PATH_FOLDERS:
        return web.json_response({"error": "Unsupported model folder"}, status=404)

    return web.json_response({"folder": folder, "paths": build_model_path_map(folder)})
